src/foam/dao/index/LRUCachingIndex.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.dao.index;
+
+public class LRUCachingIndex
+    extends ProxyIndex
+{
+  protected int size_ = 0;
+  protected final int maxSize_;
+
+  protected LRUCachingState head_ = new LRUCachingState();
+  protected LRUCachingState tail_ = new LRUCachingState();
+
+  public LRUCachingIndex(int maxSize, Index index) {
+    setDelegate(index);
+    this.head_.setNext(tail_);
+    this.tail_.setPrev(head_);
+    this.maxSize_ = maxSize;
+  }
+
+  @Override
+  public Object wrap(Object state) {
+    LRUCachingState cache = new LRUCachingState(getDelegate().wrap(state), state);
+    cache.setNext(head_.getNext());
+    cache.setPrev(head_);
+    head_.setNext(cache);
+
+    size_ += 1;
+    // remove oldest entry
+    if ( size_ >= maxSize_ ) {
+      LRUCachingState toRemove = tail_.getPrev();
+      LRUCachingState prev = toRemove.getPrev();
+      prev.setNext(tail_);
+      tail_.setPrev(prev);
+      toRemove.setValue(null);
+      size_ -= 1;
```

Comment:
You could remove this line and just add the size++ above to the else.

Reply:
Thanks. Done